Crazypilot's Space Ruins Port

Crazylemon64 opened the pull request:
Fixes up some things on #4578 - namely the area problem, atmos init bugs, badly-placed APCs, and pipes falling apart on map load.

Credits to @CrAzYPiLoT-SS13 for the original PR

Pinging @Krausus Since they found the bugs I patched.

From here on below are Crazypilot's words.

No longer a WIP PR, this is completely done.
Thanks to @KorPhaeron for this.

PRs used:
tgstation/tgstation#17939
tgstation/tgstation#17478
tgstation/tgstation#17366
tgstation/tgstation#16072
tgstation/tgstation#16022
tgstation/tgstation#13802
tgstation/tgstation#19002

Thanks to all the contributors to these PRs! Thanks to @RemieRichards for helping me resolve the runtime!

:cl: CrAzYPiLoT
rscadd: Ports randomized space ruins from /tg/ station. There is a number of maps which can be spawned at any location on the empty z-level at roundstart, and up to three random ones will be chosen each round. Credits to KorPhaeron.
rscadd: Lays the groundwork for a future lavaland implementation, if we choose to have it.
rscadd: Adds a feature to the map loader where areas with the right var set can be distinctly initialized on map load - meaning separate APCs, and the like.
rscadd: Using a z-level specific initialization freeze system, large elaborate maps like the cyberiad can be loaded without a single runtime, with minimal impact on the round elsewhere.
rscadd: Adds a delete mode to the "Fill" buildmode - just alt-click for the second corner, and everything within will be removed.
rscadd: Adds a "Jump To" function in VV, which is useful for getting an idea of what exactly you're looking at.
/:cl:

CrAzYPiLoT-SS13 commented:
You should change the name in the CL to your own though, you did most of the important stuff, I just copied and modified code from TG. Of course, I would like it if you added my name to the body of the CL, just for the record ;) Good job on this!

Crazylemon64 commented on code/modules/awaymissions/maploader/reader.dm:
```diff
@@ -349,11 +399,34 @@ var/global/dmm_suite/preloader/_preloader = new
+
+	return area_list[A]
+
 /area/template_noop
```
I'll highlight this since it might be interesting to -tg- devs - or they might realize this would just cause yet another slew of problems. What this does, is if the `/area` has `there_can_be_many` set to true, is create a distinct area for a loaded map, so that multiple space ruins maps can use the same `/area/ruin/unpowered` in their file, and not have APCs and air alarms behave linked together.

Crazylemon64 commented on code/modules/mob/living/simple_animal/hostile/mining_mobs.dm:
```diff
@@ -170,7 +171,7 @@
 /mob/living/simple_animal/hostile/asteroid/goldgrub/proc/Burrow()//Begin the chase to kill the goldgrub in time
-	if(!alerted)
+	if(!alerted && will_burrow)
 		alerted = 1
```
If you're wondering what this is about: The abandoned zoo has a goldgrub with `will_burrow` set to false - but since that var doesn't exist, attempting to load it will runtime. What this does to remedy this, is to set a var that does what you'd expect `will_burrow` to do - meaning that the goldgrub in the abandoned zoo won't be able to burrow away.

Crazylemon64 commented on _maps/map_files/RandomRuins/SpaceRuins/gasthelizards.dmm:
```diff
@@ -13,7 +13,7 @@
 "o" = (/obj/structure/closet/crate/freezer,/obj/item/weapon/reagent_containers/food/snacks/lizardmeat,/turf/simulated/floor/plasteel{icon_state = "dark"},/area/ruin/unpowered)
-"p" = (/obj/machinery/power/apc{name = "Worn-out APC"; dir = 1; pixel_x = 1; pixel_y = 26; cell_type = 2500},/turf/simulated/wall,/area/ruin/unpowered)
+"p" = (/obj/machinery/power/apc{dir = 8; environ = 0; equipment = 0; lighting = 0; locked = 0; name = "Worn-out APC"; pixel_x = -24; pixel_y = 0},/obj/structure/window/reinforced{dir = 1; layer = 2.9},/turf/simulated/floor/plasteel{dir = 1; icon_state = "blue"},/area/ruin/unpowered)
 "q" = (/obj/structure/window/reinforced{dir = 1; layer = 2.9},/turf/simulated/floor/plasteel{dir = 1; icon_state = "blue"},/area/ruin/unpowered)
```
This might have some gameplay effects: Should I modify the other vars to ensure that the APC is the same as before?

Krausus commented:
Worn-out APCs should probably just be a new subtype, so they don't need a bunch of var overrides.

I'm not sure how I feel about having to change pipe code to make map loading work, seeing as it worked fine before porting the updated maploader. That doesn't seem right.

It looks like away missions load fine now... but, out of morbid curiosity, I decided to see what happens if you *really* push the map loader to its limits...

... so, before round-start, I maploaded the Cyberiad onto an empty Z6.

Most areas look mostly fine. However, most atmos piping seems to *exist,* but has null icons. And this 4-way pipe fitting popped out of the floor for some reason.
![space station 13 2016-06-25 141719](https://cloud.githubusercontent.com/assets/10916307/16358388/b5c8e5ee-3adf-11e6-9806-06e4b6a58c34.png)
In addition to being invisible, pipes don't seem to be connecting to things like they should. Atmos seems a bit off...
![space station 13 2016-06-25 142016](https://cloud.githubusercontent.com/assets/10916307/16358409/22e4e056-3ae0-11e6-9e91-5827a0bb24ba.png)
The wiring on the solar panels looks a bit weird...
![space station 13 2016-06-25 142255](https://cloud.githubusercontent.com/assets/10916307/16358416/5a34925e-3ae0-11e6-90ac-585b94ce576d.png)
And then, there's parts of the map that seem to have loaded... wrong...
![space station 13 2016-06-25 142414](https://cloud.githubusercontent.com/assets/10916307/16358435/95292852-3ae0-11e6-9f50-e84767b1e98b.png)
"Where's the rest of the chapel floor tiles", you ask? Over here, in front of the bridge...
![space station 13 2016-06-25 142550](https://cloud.githubusercontent.com/assets/10916307/16358441/c17c263e-3ae0-11e6-82e9-08b0214ebe93.png)
Some of the solar panels snuck into the station. Also, the primate pen is cold for some reason.
![space station 13 2016-06-25 142740](https://cloud.githubusercontent.com/assets/10916307/16358461/1d098cda-3ae1-11e6-9e68-20bf16a2e007.png)
Some stuff is just floating around in space...
![space station 13 2016-06-25 143050](https://cloud.githubusercontent.com/assets/10916307/16358478/73b3e986-3ae1-11e6-9473-db78d0b58640.png)


It's weird, because a lot of the map seems to have loaded fine (aside from atmos), but there's sizable chunks that went very wrong. I wasn't expecting the map loader to load all of this perfectly, of course, but some of this is *way* off.

Crazylemon64 commented:
@Krausus The reason why we had to change the pipe code was that before, we set the turf directly, instead of going through `ChangeTurf` - `ChangeTurf` calls `updatelevels`, which calls `hide` on the pipe - which then calls `update_icons`, which will deconstruct the pipe if it hasn't been `initialize`d yet.

Krausus commented:
Alright, disregard the horrific messes I posted screenshots of previously. I mixed up Z6 and Z7, and turns out loading a template on top of existing turfs ends pretty badly. It would be kinda nice if it didn't, though.

I tried loading the Cyberiad as an away mission, and it worked almost perfectly... but atmos piping still has issues. Supply 4-way manifold fittings consistently pop out of place entirely, and when the pipenets get initialized, one of the connections on many of the 3-way pipes breaks and generates an error: "added to a pipenet while still having one (pipes leading to the same spot stacking in one turf)".

Also, the wires on catwalks (as used for the solar panels, seen above) still spawn wrong, and anything timing-sensitive (window generators, etc.) act up whenever the map loader decides to sleep. It might be best to just not have it sleep at all during object creation, and expect lag if an admin loads a particularly large template.

The fact that the map loader doesn't seem to properly load atmos piping if used after pipe init means loading any templates with piping mid-round won't work at all. That's definitely not ideal.

CrAzYPiLoT-SS13 commented:
So, uh, status?

Crazylemon64 commented:
Still need to patch the 4-way manifold, get atmos pipenets to form correctly, and then I'll investigate the weird solar panel floors.

Crazylemon64 commented:
Loading a map mid-game seems problematic, as it hits runtimes when the air controller tries to handle the turfs that haven't actually fully initialized yet.

I'll probably want something to halt all processes while a map is loading, then.

Crazylemon64 commented:
I found the answer to the manifolds problem: The manifold connection directions were being initialized before it was actually rotated, causing it not to connect as it should. After doing so, pipe connections work like a charm.

Crazylemon64 commented:
As for the weird super-red cables, that APPEARS to be a mapping issue - apparently the solar panel turfs have multiple turfs defined on their tiles: An airless plating, as well as a catwalk. These are presently out of order, so that the map-loader overlay code gets confused and places underlays on the cable. I've added a tweak that'll make it really annoyingly obvious if such a map error exists.

Crazylemon64 commented:
@Krausus Should have your concerns addressed now.

Crazylemon64 commented:
One of the last things is to get window spawners to defer their sublimation until the map is fully loaded. Should be as simple as it sounds.

Crazylemon64 commented:
The LAST thing to get this perfect would be to get the atmos system to ignore the turfs while the map is still loading - That, and wall smoothing overlays need to be perfected for mid-round maploads. But, I'll leave that for tomorrow.

tigercat2000 commented on code/game/objects/effects/spawners/windowspawner.dm:
```diff
@@ -6,7 +6,7 @@
 	var/windowtospawn = /obj/structure/window/basic
 
-/obj/effect/spawner/window/New()
+/obj/effect/spawner/window/initialize()
```
this will make it non-viable to spawn these directly and in mass

Crazylemon64 replied:
Is `initialize` not called <a href=https://github.com/ParadiseSS13/Paradise/blob/master/code/game/atoms_movable.dm#L26-L28>whenever an `obj` is spawned after round init</a>?

tigercat2000 replied:
oh
huh
i literally saw that and failed to notice what it does

Crazylemon64 commented:
@Krausus This now loads without a single runtime.

The mechanic of doing so, was moving the `air_update_turf` in many objects, such as airlocks, windows, and podlocks, to the `initialize` proc - that way, air initialization could be fully deferred until the map has fully loaded and atmos has set in.

The mechanics of deferring object initialization have been refined, as well - it is now restricted to the z-levels maps are being loaded on through a list of numbers, and initialization works fine on z-levels that aren't having anything loaded on them. Anything created on a z-level that has something being loaded at the same time, provided it isn't created within the bounds of the loaded map, however, will fail to initialize, if something else isn't handling its initialization for it.

Unfortunately, right now the map loader is not re-entrant, as a result of these deferred initializations, but then again BYOND's multi-threading support isn't that substantial either.

The last thing needed, is to set all the correct vars on the map templates provided with the code here.

Crazylemon64 commented:
It seems some tiles on the moon outpost have burnt plating in the same spot as they do cables. Since burnt tiles can't have cabling, I think they lose the cables when the map is loaded. I'll chalk this up as a "feature" and call it a day.

Crazylemon64 commented:
Ready for review. @TheDZD @Fox-McCloud

Krausus commented:
> Unfortunately, right now the map loader is not re-entrant

That's going to be a problem. #4700 uses map template loaders to randomize which rooms get loaded in, which means there will be a bunch of templates being loaded simultaneously. Allowing multiple map loaders to run at the same time is kinda important.

> Anything created on a z-level that has something being loaded at the same time, provided it isn't created within the bounds of the loaded map, however, will fail to initialize, if something else isn't handling its initialization for it.

This will probably cause problems if a template is loaded onto the station Z-level, won't it? That's not good. I'm guessing this will also cause unpredictable problems if a template gets loaded on top of existing objects.

> It seems some tiles on the moon outpost have burnt plating in the same spot as they do cables. Since burnt tiles can't have cabling, I think they lose the cables when the map is loaded. I'll chalk this up as a "feature" and call it a day.

But... but they **can** have cabling. It exists in the map as loaded by the current maploader, and you can spawn in and manually place down new cabling where it's missing from this loader.  :confused:

tigercat2000 commented:
the maploader is used for admin map templates as well, which are arbitrary files that can be placed literally anywhere on the map

Crazylemon64 commented:
It's re-entrant now, and making it defer initializations is now an optional thing set to off by default

Also the stupid cables now stay on the burnt plating

Crazylemon64 commented:
Alright, so the map loader no longer defers initialization on its z-level by default - that requires an additional argument, and should be used when you want to load something elaborate like the cyberiad, or some other atmos-containing station. The admin template loading prompt now allows admins the choice to defer init or not, with all the consequences that entails. 

If the caller chooses to delay initialization on the z-level work is being done on, all other z-levels should be just fine, and I think the only things with an actual `initialize` proc are structures that obstruct atmos (windows, podlocks, that stuff) as well as machinery.

Crazylemon64 commented:
Ready for review yet again @tigercat2000 @Krausus @TheDZD @Fox-McCloud

tigercat2000 commented on code/game/atoms_movable.dm:
```diff
@@ -20,10 +20,25 @@
+	// If you're wondering what goofery this is, this is for things that need the environment
+	// around them set up - like `air_update_turf` and the like
+	if(!auto_init)
+		//no-op
```
do not make no-op statements please
use early returns if you must

Crazylemon64 replied:
Should I condense this into an impenetrable massive single if conditional that's repeated a whole bunch

TheDZD commented on .gitignore:
```diff
@@ -7,7 +7,7 @@
 #ignore any files in config/, except those in subdirectories.
 /config/*
-!/config/*/*
+!config/**/*/
```
What does this do/why is it needed exactly?

Crazylemon64 replied:
This seems to be a leftover from the PR I took it from - I'll restore it to what it was previously.

CrAzYPiLoT-SS13 replied:
No, don't. I added that because it wasn't properly set up before. Now Git
won't ignore changes done to the example folder, like it should, and will
just ignore changes in the root config folder. That way we can add and
change config files.

Krausus replied:
But that's already what this line does. Look at the history of the files in the example folder, and you'll see they've been changed plenty of times. This change changes nothing...

CrAzYPiLoT-SS13 replied:
My best guess is that Git recently updated the syntax for subfolders, cause it wouldn't work in my PR. You can try to revert it, if the config file changes don't get ignored, you can push the change. Although, in that case, I would wonder why it wouldn't work on my system.

TheDZD commented on code/ATMOSPHERICS/datum_pipeline.dm:
```diff
@@ -47,6 +48,7 @@ var/pipenetwarnings = 10
 		addMachineryMember(base)
 	if(!air)
 		air = new
+//	log_startup_progress("Making pipenet [pipenets_made++]")
```
Leftover debug code?

TheDZD commented on code/controllers/master_controller.dm:
```diff
@@ -11,16 +11,16 @@ var/global/air_processing_killed = 0
 /datum/controller
-  var/processing = 0
-  var/iteration = 0
-  var/processing_interval = 0
```
Oh god why were these using spaces for indentation?

Crazylemon64 commented:
I found a bug in the map loader that was making Space Hotel cry - if it hits a map template and attempts to load another map, it'll reset the position of the reader once it's finished parsing the block, and re-load the entire map - resulting in limbo. I've patched this, and will now send a PR over to -tg- to fix it on their side too.

Krausus commented:
Sorry, been too distracted to take another look at the state of this.

It sounds like the initialization stuff is turning out to be a mess. I'm curious - instead of having stuff skip initializing, then iterating blindly through everything in the world/z-level/block that *might* need to initialize... what if you just had a global list, everything that deferred init added itself to it, and then the list gets chewed through when it's safe to initialize?

That way, you should be able to guarantee you only init the stuff that needs it, and don't happen to miss anything that spawns elsewhere in the same Z-level during loading.

Crazylemon64 commented:
I DID have an idea that worked similarly, however - some things, like atmos, need to be set up before initialization, and that's handled outside of loading the map itself - in `late_setup_level`, which is called by whatever is using the map loader. Though - if I moved the z-level dirt system outside of the map loader itself, and out to what calls it (currently only the map templates, and the away mission setup), I should be able to get things working just fine.

Anyways, right now I have `delay_init` as an option to the map loader, but now that I've mulled it over I think it might be wise, as you suggest, to move init delaying out to the responsibility of the caller. That way, it can all be handled nicely without permanently freezing anything that shouldn't be.

Also I was a snarky butt to @Krausus in the first iteration of the cable fix and I apologize for being one.

Crazylemon64 commented:
Alright, I implemented @Krausus' idea - the loader should work like a dream now, provided the map it loads onto is empty before you do so.

We might want a system to efficiently wipe out the area the map is being loaded to, but there's no good way of telling where needs erasing without loading the map to the point where you're already putting turfs down. Well, MAYBE if we did something that searched for all the points where there were no-op turfs, left those spots alone, and flattened everything else - but that's a matter for another PR. This stands on its own as it is.

Crazylemon64 commented:
Oh. Pipes don't use the normal init path.

Bother.

TheDZD commented:
Conflict.

tigercat2000 commented on code/game/atoms_movable.dm:
```diff
@@ -36,6 +39,16 @@
+// at which point object creations are a fair toss more seldom
+/atom/movable/proc/attempt_init()
+	var/turf/T = get_turf(src)
+	if (T && zlevels.is_zlevel_dirty(T.z))
```
no space between `if` and `(`

Crazylemon64 replied:
Thanks for the reminder that the merge conflict wouldn't have given me

tigercat2000 commented on code/game/objects/structures/mineral_doors.dm:
```diff
@@ -16,152 +16,155 @@
+
+/obj/structure/mineral_door/proc/Dismantle(devastated = 0)
+	if(!devastated)
+		if (mineralType == "metal")
```
no space between `if` and `(`

tigercat2000 commented on code/game/objects/structures/mineral_doors.dm:
```diff
@@ -16,152 +16,155 @@
+			for(var/i = 1, i <= oreAmount, i++)
+				new ore(get_turf(src))
+	else
+		if (mineralType == "metal")
```
no space between `if` and `(`

tigercat2000 commented on code/modules/awaymissions/zlevel.dm:
```diff
@@ -1,35 +1,110 @@
+		//map_transition_config.Add(AWAY_MISSION_LIST)
+
+		for(var/obj/effect/landmark/L in landmarks_list)
+			if (L.name != "awaystart")
```
no space between `if` and `(`

tigercat2000 commented on code/modules/awaymissions/zlevel.dm:
```diff
@@ -1,35 +1,110 @@
+			//map_transition_config.Add(AWAY_MISSION_LIST)
+
+			for(var/obj/effect/landmark/L in landmarks_list)
+				if (L.name != "awaystart")
```
no space between `if` and `(`

tigercat2000 commented on code/modules/spacial_allocator/zlevel_manager.dm:
```diff
@@ -0,0 +1,89 @@
+/datum/zlev_manager/proc/get_zlev(z)
+  if(z < 1)
+    throw EXCEPTION("Non-positive z level given!")
+  else if (z > z_list.len)
```
no space between `if` and `(`

Crazylemon64 commented:
Alright, code style politics out of the way.
I've also made sure that the late init system works with elaborate maps such as space hotel, or the cyberiad. Pipes and meters init, as well as everything else, and I think I also fixed a bug with area code that set `power_light` to 0 on areas with `requires_power` set to 0.
Review away.
@Krausus @tigercat2000 @TheDZD @Fox-McCloud

Crazylemon64 commented:
If it isn't obvious, the merge conflict's been resolved.

Fox-McCloud commented on code/modules/spacial_allocator/zlevel.dm:
```diff
@@ -0,0 +1,73 @@
+  zpos = z
+
+/datum/zlevel/proc/resume_init()
+  if(dirt_count > 0)
```
the spacing on this seems wrong.

Crazylemon64 replied:
In what manner? The spacing on the greater than sign?

Fox-McCloud commented on code/_globalvars/mapping.dm:
```diff
@@ -54,3 +54,10 @@ var/list/carplist = list() //list of all carp-spawn landmarks
+
+var/list/datum/map_template/ruins_templates = list()
+var/list/datum/map_template/space_ruins_templates = list()
+//var/list/datum/map_template/lava_ruins_templates = list()
```
:eyes:

CrAzYPiLoT-SS13 replied:
I added that for future lavaland implementations (maybe?) It will be easier to just uncomment some of the code that is already in, than to make a port for it all.

Fox-McCloud replied:
it's fine, doesn't need to be removed, just making eyes at it 'cause I can.

Fox-McCloud commented on code/datums/helper_datums/map_template.dm:
```diff
@@ -9,59 +9,153 @@
+		ruins_templates[R.name] = R
+
+		/*
+		if(istype(R, /datum/map_template/ruin/lavaland))
```
:eyes:

CrAzYPiLoT-SS13 replied:
Past comment

Fox-McCloud replied:
it's fineeeee.

Fox-McCloud commented on code/modules/awaymissions/corpse.dm:
```diff
@@ -285,3 +288,10 @@
+/obj/effect/landmark/corpse/abductor //Connected to ruins, for some reason?
+	name = "abductor"
+	mobname = "???"
+	mob_species = /datum/species/abductor
```
should be a text string, it's what `set_species` expects

Fox-McCloud commented:
Alright, question. How does this handle when you load a template onto an existing map? Say like plopping one of these ruins down in the middle of the mining asteroid?

CrAzYPiLoT-SS13 commented:
From what I understand from the conversation, it works fine, by replacing all the turfs. Don't take my word for it though. Better ping @Crazylemon64

Crazylemon64 commented:
If you load it on top of an existing map, it will leave all items already present intact, but replace all turfs - I fixed a bug with the map loader that failed to overwrite plating icons if they were of the same path.

If you try to load a map that contains atmos (cyberiad, metastation, etc.) on top of another map, atmos runtimes will arise as a result of the air controller trying to process turfs before it's done - I've added procs to empty out an area, though right now they leave residue if it hits something like a closet.

I'd have the map loader empty tiles before it loads, but I don't want to cause an unnecessary performance hit if I don't have to - so I've instead opted to have a proc to empty out a region of space - `empty_region` and `empty_rect`.

For admins that want to load over a section of space, I've added a function to the "Fill" buildmode, where if the second click is done while Alt is held, it will instead erase the entire section of space.

Procs that desire to place a map template down where there might possibly be turfs can instead call `empty region`.

However, the purpose that @Fox-McCloud mentions (loading ruins into the asteroid) SHOULD work OK, as it should be fast enough to complete before the atmos controller catches up and yells at it.

Fox-McCloud commented:
Hmm, bit worried about post runtime loading of templates causing issues.

Reason I mention this, specifically is the shelter map template that capsules use--these capsules originally didn't use templates (when I considered porting them), but now they do---just a bit worried at how that will behave if you were to use it in the middle of the station or anywhere else, for that matter.

Crazylemon64 commented:
The shelters are small enough that it'll be done LOONG before the atmos controller can complain.

Granted, I should probably find some way of suspending adding turfs to the atmos controller before things are done.

Crazylemon64 commented:
Eh, the air controller is so deeply integrated into so many systems of the game that I don't think I can find a clean way to suspend atmos init. Not without unintended consequences, at least.
